ethcore/src/evm/wasm/runtime.rs

```diff
@@ -0,0 +1,356 @@
+	) -> Result<Option<interpreter::RuntimeValue>, interpreter::Error> {
+		Ok(None)
+	}
+
+	/// Write call descriptor to wasm memory
+	pub fn write_descriptor(&mut self, call_args: CallArgs) -> Result<WasmPtr, Error> {
+		let d_ptr = self.alloc(16)?;
+
+		let args_len = call_args.len();
+		let args_ptr = self.alloc(args_len)?;
+
+		// write call descriptor
+		// call descriptor is [args_ptr, args_len, return_ptr, return_len]
+		//   all are 4 byte length, last 2 are zeroed
+		let mut d_buf = [0u8; 16];
+		LittleEndian::write_u32(&mut d_buf[0..4], args_ptr);
+		LittleEndian::write_u32(&mut d_buf[4..8], args_len);
+		self.memory.set(d_ptr, &d_buf)?;
+
+		// write call args to memory
+		self.memory.set(args_ptr, &call_args.address)?;
+		self.memory.set(args_ptr+20, &call_args.sender)?;
+		self.memory.set(args_ptr+40, &call_args.origin)?;
+		self.memory.set(args_ptr+60, &call_args.value)?;
+		self.memory.set(args_ptr+92, &call_args.data)?;
+		
+		Ok(d_ptr.into())
+	}
+
+	fn debug_log(&mut self, context: interpreter::CallerContext) 
+			-> Result<Option<interpreter::RuntimeValue>, interpreter::Error> 
+	{
+		let msg_len = context.value_stack.pop_as::<i32>()? as u32;
+		let msg_ptr = context.value_stack.pop_as::<i32>()? as u32;
+		
+		let msg = String::from_utf8(self.memory.get(msg_ptr, msg_len as usize)?)
+			.map_err(|_| interpreter::Error::Trap("Debug log utf-8 decoding error".to_owned()))?;
+
+		trace!(target: "wasm", "Contract debug message: {}", msg);
+
+		Ok(None)
+	}
+
+	/// Query current gas left for execution
+	pub fn gas_left(&self) -> Result<u64, Error> {
+		if self.gas_counter > self.gas_limit { return Err(Error::InvalidGasState); }
+		Ok(self.gas_limit - self.gas_counter)
+	}
+
+	/// Shared memory reference
+	pub fn memory(&self) -> &interpreter::MemoryInstance {
+		&*self.memory
+	}
+
+	fn mem_copy(&self, context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error> 
+	{
+		let len = context.value_stack.pop_as::<i32>()? as u32;
+		let dst = context.value_stack.pop_as::<i32>()? as u32;
+		let src = context.value_stack.pop_as::<i32>()? as u32;
+
+		let mem = self.memory().get(src, len as usize)?;
+		self.memory().set(dst, &mem)?;
+
+		Ok(Some(0i32.into()))
+	}
+}
+
+impl<'a> interpreter::UserFunctionExecutor for Runtime<'a> {
+	fn execute(&mut self, name: &str, context: interpreter::CallerContext) 
+		-> Result<Option<interpreter::RuntimeValue>, interpreter::Error>
+	{
+		match name {
+			"_malloc" => {
+				self.malloc(context)
+			},
+			"_free" => {
+				// Since it is arena allocator, free does nothing
+				// todo: update if changed
+				self.user_noop(context)
+			},
+			"_storage_read" => {
+				self.storage_read(context)
+			},
+			"_storage_write" => {
+				self.storage_write(context)
+			},
+			"_suicide" => {
+				self.suicide(context)
+			},
+			"_create" => {
+				self.create(context)
+			},
+			"_debug" => {
+				self.debug_log(context)
+			},
+			"gas" => {
```

Comment:
You cannot have named functions in webassembly, only imports or exports are named
And if C program will import `gas` function explicitly, it will be invalid as a contract and will fail instantly, before even running (as wasm module validation fails when there are duplicated imports, there should be rule about it in webassembly official testsuite)